### PR TITLE
cpu/stm32/periph/dma: add support for STM32F3

### PR DESCRIPTION
**Pull request by hugueslarrive**
### Contribution description
Add support for STM32F3.

### Testing procedure
Only tested on STM32F334r8.

### Issues/PRs references
Moved from #14734.

### CODE REVIEW

File: cpu/stm32/periph/dma.c [hunk 1]
```diff
@@ -196,6 +192,7 @@ static IRQn_Type dma_get_irqn(int stream)
     if (stream < 7) {
         return ((IRQn_Type)((int)DMA1_Channel1_IRQn + stream));
     }
+#if !defined(CPU_FAM_STM32F3)
 #if defined(CPU_FAM_STM32F1)
```
Comment (benpicco) [inline cpu/stm32/periph/dma.c:196]: Seems redundant. If `CPU_FAM_STM32F1` is defined, `CPU_FAM_STM32F3` won't be defined and vice-versa.

Comment (hugueslarrive) [reply cpu/stm32/periph/dma.c:196]: There is a #else to #if defined(CPU_FAM_STM32F1) which result in `error: 'DMA2_Channel1_IRQn' undeclared` at compilation time. Same thing 9 lines later. This drop the code up to line 212.

**Comment by hugueslarrive** [comment]
Tried with the bluepill (f103c8) and dma_stop lines seems to be also needed for this one although not in the f3 family so I think it's safer to always do it as it shouldn't be a problem for MCUs that don't need it.

File: cpu/stm32/periph/dma.c (hunk 1 above)
Comment (benpicco) [reply cpu/stm32/periph/dma.c:196]: Ah ok - better add comments to the `#endif` so it's apparent what the matching `#if` was.

Comment (hugueslarrive) [reply cpu/stm32/periph/dma.c:196]: Comment added.

File: cpu/stm32/periph/spi.c [hunk 2]
```diff
@@ -256,8 +256,9 @@ static void _transfer_dma(spi_t bus, const void *out, void *in, size_t len)
     dma_wait(spi_config[bus].rx_dma);
     dma_wait(spi_config[bus].tx_dma);
 
-    /* No need to stop the DMA here, it is automatically disabled when the
-     * transfer is finished, only wait for SPI to leave the busy state */
+    dma_stop(spi_config[bus].rx_dma);
+    dma_stop(spi_config[bus].tx_dma);
```
Comment (fjmolinas) [inline cpu/stm32/periph/spi.c:260]: @bergzand you worked with DMA optimization a bunch I think, is this worth optimizing for platforms who do or do not need to use the stop?

**Comment by fjmolinas** [comment]
> Tried with the bluepill (f103c8) and dma_stop lines seems to be also needed for this one although not in the f3 family so I think it's safer to always do it as it shouldn't be a problem for MCUs that don't need it.

How did you test this? Is there some test output you can post?

**Comment by hugueslarrive** [comment]
Hello @fjmolinas,

Using PR #14767 if you remove dma_stop lines and launch a bench from tests/periph_spi_dma: 
```
2020-08-18 13:34:12,087 # 
2020-08-18 13:34:43,737 # main(): This is RIOT! (Version: 2020.07-devel-2536-g24ba4a-boards/common/blxxxpill)
2020-08-18 13:34:43,740 # Manual SPI peripheral driver test
2020-08-18 13:34:43,746 # Refer to the README.md file for more information.
2020-08-18 13:34:43,746 # 
2020-08-18 13:34:43,749 # There are 2 SPI devices configured for your platform.
init 0 0 3 0 0
2020-08-18 13:35:19,473 #  init 0 0 3 0 0
2020-08-18 13:35:19,479 # SPI_DEV(0) initialized: mode: 0, clk: 3, cs_port: 0, cs_pin: 0
bench
2020-08-18 13:35:27,850 #  bench
2020-08-18 13:35:27,853 # ### Running some benchmarks, all values in [us] ###
2020-08-18 13:35:27,856 # ### Test				Transfer time	user time
2020-08-18 13:35:27,856 # 
2020-08-18 13:35:41,371 # main(): This is RIOT! (Version: 2020.07-devel-2536-g24ba4a-boards/common/blxxxpill)
2020-08-18 13:35:41,374 # Manual SPI peripheral driver test
2020-08-18 13:35:41,379 # Refer to the README.md file for more information.
2020-08-18 13:35:41,380 # 
2020-08-18 13:35:41,383 # There are 2 SPI devices configured for your platform.
```
As you can see there is no output after more than 10 seconds (I pressed reset button at 13:35:41).

With those lines it works:
```
2020-08-18 13:48:04,910 # main(): This is RIOT! (Version: 2020.07-devel-2536-g24ba4a-boards/common/blxxxpill)
2020-08-18 13:48:04,913 # Manual SPI peripheral driver test
2020-08-18 13:48:04,918 # Refer to the README.md file for more information.
2020-08-18 13:48:04,919 # 
2020-08-18 13:48:04,922 # There are 2 SPI devices configured for your platform.
init 0 0 3 0 0
2020-08-18 13:48:08,336 #  init 0 0 3 0 0
2020-08-18 13:48:08,342 # SPI_DEV(0) initialized: mode: 0, clk: 3, cs_port: 0, cs_pin: 0
bench
2020-08-18 13:48:13,037 #  bench
2020-08-18 13:48:13,041 # ### Running some benchmarks, all values in [us] ###
2020-08-18 13:48:13,044 # ### Test				Transfer time	user time
2020-08-18 13:48:13,044 # 
2020-08-18 13:48:13,062 #  1 - write 1000 times 1 byte:			13266	13270
2020-08-18 13:48:13,084 #  2 - write 1000 times 2 byte:			18718	18722
2020-08-18 13:48:13,285 #  3 - write 1000 times 100 byte:		195545	15451
2020-08-18 13:48:13,316 #  4 - write 1000 times 1 byte to register:	26134	26138
2020-08-18 13:48:13,352 #  5 - write 1000 times 2 byte to register:	31746	31750
2020-08-18 13:48:13,566 #  6 - write 1000 times 100 byte to register:	208475	28404
2020-08-18 13:48:13,588 #  7 - read 1000 times 2 byte:			18599	18603
2020-08-18 13:48:13,788 #  8 - read 1000 times 100 byte:		195259	15139
2020-08-18 13:48:13,824 #  9 - read 1000 times 2 byte from register:	31689	31694
2020-08-18 13:48:14,038 # 10 - read 1000 times 100 byte from register:	208370	28291
2020-08-18 13:48:14,060 # 11 - transfer 1000 times 2 byte:		18523	18528
2020-08-18 13:48:14,260 # 12 - transfer 1000 times 100 byte:		195279	15197
2020-08-18 13:48:14,297 # 13 - transfer 1000 times 2 byte to register:	31551	31555
2020-08-18 13:48:14,510 # 14 - transfer 1000 times 100 byte to register:208278	28182
2020-08-18 13:48:14,531 # 15 - acquire/release 1000 times:		17363	17368
2020-08-18 13:48:15,532 # -- - SUM:					1418795	338292
2020-08-18 13:48:15,533 # 
2020-08-18 13:48:15,536 # ### All runs complete ###
```

File: cpu/stm32/periph/spi.c (hunk 2 above)
Comment (bergzand) [reply cpu/stm32/periph/spi.c:260]: I think it is worth optimizing it a bit, but definitely out of scope here. I agree with bringing back `dma_stop()` here.

Comment (hugueslarrive) [reply cpu/stm32/periph/spi.c:260]: Test on f746zg without dma_stop:
```
2020-08-18 13:53:08,734 #  init 0 0 4 0 0
2020-08-18 13:53:08,740 # SPI_DEV(0) initialized: mode: 0, clk: 4, cs_port: 0, cs_pin: 0
bench
2020-08-18 13:53:14,653 #  bench
2020-08-18 13:53:14,658 # ### Running some benchmarks, all values in [us] ###
2020-08-18 13:53:14,661 # ### Test				Transfer time	user time
2020-08-18 13:53:14,661 # 
2020-08-18 13:53:14,672 #  1 - write 1000 times 1 byte:			7159	7162
2020-08-18 13:53:14,688 #  2 - write 1000 times 2 byte:			11913	11915
2020-08-18 13:53:14,821 #  3 - write 1000 times 100 byte:		129408	9782
2020-08-18 13:53:14,841 #  4 - write 1000 times 1 byte to register:	14693	14695
2020-08-18 13:53:14,865 #  5 - write 1000 times 2 byte to register:	19945	19949
2020-08-18 13:53:15,007 #  6 - write 1000 times 100 byte to register:	136951	17615
2020-08-18 13:53:15,023 #  7 - read 1000 times 2 byte:			11962	11964
2020-08-18 13:53:15,156 #  8 - read 1000 times 100 byte:		129075	9449
2020-08-18 13:53:15,181 #  9 - read 1000 times 2 byte from register:	20302	20305
2020-08-18 13:53:15,323 # 10 - read 1000 times 100 byte from register:	137127	17879
2020-08-18 13:53:15,339 # 11 - transfer 1000 times 2 byte:		11762	11764
2020-08-18 13:53:15,473 # 12 - transfer 1000 times 100 byte:		129093	9484
2020-08-18 13:53:15,498 # 13 - transfer 1000 times 2 byte to register:	19927	19930
2020-08-18 13:53:15,640 # 14 - transfer 1000 times 100 byte to register:136965	17634
2020-08-18 13:53:15,654 # 15 - acquire/release 1000 times:		10081	10084
2020-08-18 13:53:16,657 # -- - SUM:					926363	209611
2020-08-18 13:53:16,657 # 
2020-08-18 13:53:16,659 # ### All runs complete ###
```
Then with dma_stop:
```
2020-08-18 13:54:02,402 # init 0 0 4 0 0
2020-08-18 13:54:02,407 # SPI_DEV(0) initialized: mode: 0, clk: 4, cs_port: 0, cs_pin: 0
bench
2020-08-18 13:54:11,806 #  bench
2020-08-18 13:54:11,811 # ### Running some benchmarks, all values in [us] ###
2020-08-18 13:54:11,814 # ### Test				Transfer time	user time
2020-08-18 13:54:11,814 # 
2020-08-18 13:54:11,826 #  1 - write 1000 times 1 byte:			8447	8449
2020-08-18 13:54:11,843 #  2 - write 1000 times 2 byte:			12743	12746
2020-08-18 13:54:11,976 #  3 - write 1000 times 100 byte:		130242	10707
2020-08-18 13:54:11,998 #  4 - write 1000 times 1 byte to register:	16589	16592
2020-08-18 13:54:12,024 #  5 - write 1000 times 2 byte to register:	22131	22133
2020-08-18 13:54:12,169 #  6 - write 1000 times 100 byte to register:	138918	19422
2020-08-18 13:54:12,186 #  7 - read 1000 times 2 byte:			13322	13326
2020-08-18 13:54:12,320 #  8 - read 1000 times 100 byte:		129964	10448
2020-08-18 13:54:12,347 #  9 - read 1000 times 2 byte from register:	22408	22411
2020-08-18 13:54:12,492 # 10 - read 1000 times 100 byte from register:	139159	19643
2020-08-18 13:54:12,508 # 11 - transfer 1000 times 2 byte:		12638	12641
2020-08-18 13:54:12,644 # 12 - transfer 1000 times 100 byte:		129992	10435
2020-08-18 13:54:12,669 # 13 - transfer 1000 times 2 byte to register:	22103	22106
2020-08-18 13:54:12,814 # 14 - transfer 1000 times 100 byte to register:138937	19410
2020-08-18 13:54:12,827 # 15 - acquire/release 1000 times:		10067	10070
2020-08-18 13:54:13,831 # -- - SUM:					947660	230539
2020-08-18 13:54:13,831 # 
2020-08-18 13:54:13,833 # ### All runs complete ###
```
The impact seems significant to me so I have investigated further.

The EN bit description of the devices with SxCR register says that `This bit may be cleared by hardware:
–
on a DMA end of transfer (stream ready to be configured)` unlike those with CCR register so I think that only those with DMA_CCR_EN defined need those lines, not those with DMA_SxCR_EN.

**Comment by hugueslarrive** [comment]
Tested on the f103 and f746.

File: cpu/stm32/periph/spi.c (hunk 2 above)
Comment (hugueslarrive) [reply cpu/stm32/periph/spi.c]: @bergzand Your opinion ?

File: cpu/stm32/periph/dma.c [hunk 3]
```diff
@@ -51,7 +47,7 @@
 #define RCC_MASK_DMA2           RCC_AHB1ENR_DMA2EN
 #else /* CPU_FAM_STM32L4 */
 #define CLOCK                   AHB
-#if CPU_FAM_STM32F1 || CPU_FAM_STM32L1
+#if CPU_FAM_STM32F1 || CPU_FAM_STM32L1 || CPU_FAM_STM32F3
```
Comment (fjmolinas) [inline cpu/stm32/periph/dma.c:50]: Keep ordering?

```suggestion
#if CPU_FAM_STM32F1  || CPU_FAM_STM32F3 || CPU_FAM_STM32L1
```

**Review by fjmolinas — COMMENTED** [review]
Just on nit-pick, please squash all commits into one `cpu/stm32/periph/dma: add support for STM32F3` and a second commit that enables `dma_stop` for all cpu's that need it. Maybe the later can be the first commit , maybe something like: `cpu/stm32/periph/spi: use dma_stop for STM32s that need it`

**Review by fjmolinas — APPROVED** [review]
ACK.

File: cpu/stm32/periph/spi.c [hunk 4]
```diff
@@ -256,8 +256,10 @@ static void _transfer_dma(spi_t bus, const void *out, void *in, size_t len)
     dma_wait(spi_config[bus].rx_dma);
     dma_wait(spi_config[bus].tx_dma);
 
-    /* No need to stop the DMA here, it is automatically disabled when the
-     * transfer is finished, only wait for SPI to leave the busy state */
+#ifdef DMA_CCR_EN
```
Comment (fjmolinas) [inline cpu/stm32/periph/spi.c:259]: Travis is pointing out a trailing whitespace here.

Comment (hugueslarrive) [reply cpu/stm32/periph/spi.c]: Should be ok now

**Comment by fjmolinas** [comment]
GO! Thanks for the contribution @hugueslarrive!

**Pull request merged**